crates/task-impls/src/consensus/proposal.rs

```diff
@@ -226,8 +232,338 @@ pub async fn create_and_send_proposal<TYPES: NodeType>(
+/// standard case proposal scenario.
+#[allow(clippy::too_many_arguments)]
+async fn publish_proposal_from_commitment_and_metadata<TYPES: NodeType>(
+    cur_view: TYPES::Time,
+    view: TYPES::Time,
+    sender: Sender<Arc<HotShotEvent<TYPES>>>,
+    quorum_membership: Arc<TYPES::Membership>,
+    public_key: TYPES::SignatureKey,
+    private_key: <TYPES::SignatureKey as SignatureKey>::PrivateKey,
+    consensus: Arc<RwLock<Consensus<TYPES>>>,
+    commitment_and_metadata: CommitmentAndMetadata<TYPES>,
+    delay: u64,
+    formed_upgrade_certificate: Option<UpgradeCertificate<TYPES>>,
+    decided_upgrade_cert: Option<UpgradeCertificate<TYPES>>,
+    proposal_cert: Option<ViewChangeEvidence<TYPES>>,
+) -> Result<JoinHandle<()>> {
+    let (parent_leaf, state) = get_parent_leaf_and_state(
+        cur_view,
+        view,
+        quorum_membership,
+        public_key.clone(),
+        consensus.clone(),
+    )
+    .await?;
+
+    // In order of priority, we should try to attach:
+    //   - the parent certificate if it exists, or
+    //   - our own certificate that we formed.
+    // In either case, we need to ensure that the certificate is still relevant.
+    //
+    // Note: once we reach a point of potentially propose with our formed upgrade certificate, we will ALWAYS drop it. If we cannot immediately use it for whatever reason, we choose to discard it.
+    // It is possible that multiple nodes form separate upgrade certificates for the some upgrade if we are not careful about voting. But this shouldn't bother us: the first leader to propose is the one whose certificate will be used. And if that fails to reach a decide for whatever reason, we may lose our own certificate, but something will likely have gone wrong there anyway.
+    let mut proposal_upgrade_certificate = parent_leaf
+        .get_upgrade_certificate()
+        .or(formed_upgrade_certificate);
+
+    if !proposal_upgrade_certificate
+        .clone()
+        .is_some_and(|cert| cert.is_relevant(view, decided_upgrade_cert.clone()).is_ok())
+    {
+        proposal_upgrade_certificate = None;
+    }
+
+    // We only want to proposal to be attached if any of them are valid.
+    let proposal_certificate = proposal_cert
+        .as_ref()
+        .filter(|cert| cert.is_valid_for_view(&view))
+        .cloned();
+
+    Ok(async_spawn(async move {
+        create_and_send_proposal(
+            public_key,
+            private_key,
+            consensus,
+            sender,
+            view,
+            commitment_and_metadata,
+            parent_leaf.clone(),
+            state,
+            proposal_upgrade_certificate,
+            proposal_certificate,
+            delay,
+        )
+        .await;
+    }))
+}
+
+/// Publishes a proposal if there exists a value which we can propose from. Specifically, we must have either
+/// `commitment_and_metadata`, or a `decided_upgrade_cert`.
+#[cfg(feature = "dependency-tasks")]
```

Comment:
Do we still have to have two `publish_proposal_if_able` functions? Now that this is a standalone function, I think we can just use the `cfg` feature to gate whether we want to call this function.

I don't think having two implementations causes any bug, but there may be confusion, since the function name suggests we would propose if able, but it actually won't when compiling with `dependency-tasks`.

Reply:
Yeah the issue is just that the call site will still call the function, so if we don't gate it with an equivalent function it will not compile.

Reply:
Can we not gate `publish_proposal_if_able`, but the line of code that calls `publish_proposal_if_able` in each task?

Reply:
I made a helper in the consensus task to bridge the gap and now I've localized the change to being just there. Latest push has it.